### PR TITLE
fixed StackOverflowException in TemplateRegion #53

### PR DESCRIPTION
**Pull request by KartonDCP**
This problem occurs due to the `_jmixEntityEntry` field that appears in runtime in `TemplateTableDescription` instance because it annotated with `JmixEntity`. This metadata field should not be serialized. In this way `gson` should ignore any metadata-inserted fields.

### CODE REVIEW

File: reports-ui/src/main/java/io/jmix/reportsui/wizard/template/generators/TableGenerator.java [hunk 1]
```diff
@@ -16,22 +16,32 @@
...
 import io.jmix.reportsui.wizard.template.Generator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
 
 import java.util.LinkedList;
 import java.util.List;
 
+@Component("reports_TableGenerator")
+@Scope("prototype")
```
Comment (NikitaShchienko) [inline reports-ui/src/main/java/io/jmix/reportsui/wizard/template/generators/TableGenerator.java:36]: Replace with  BeanDefinition.SCOPE_PROTOTYPE

**Review by NikitaShchienko — APPROVED** [review]

File: reports-ui/src/main/java/io/jmix/reportsui/wizard/template/TemplateGeneratorImpl.java [hunk 2]
```diff
@@ -20,11 +20,12 @@
 import io.jmix.reports.entity.wizard.TemplateFileType;
 import io.jmix.reports.exception.TemplateGenerationException;
 import io.jmix.reportsui.wizard.template.generators.*;
+import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.context.annotation.Scope;
 import org.springframework.stereotype.Component;
 
 @Component("report_TemplateGenerator")
-@Scope("prototype")
```
Comment (KartonDCP) [inline reports-ui/src/main/java/io/jmix/reportsui/wizard/template/TemplateGeneratorImpl.java:27]: also replaced with BeanDefinition.SCOPE_PROTOTYPE

File: reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.util.Arrays;
+
+public class StrategyGsonBuilder {
+
+    protected final static String METADATA_STARTS_CHAR = "_";
+
+    public enum SerializationPolicy {
+        ignoreMetadata,
```
Comment (gorbunkov) [inline reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java:32]: Enumeration values are usually written in uppercase with underscore, e.g. IGNORE_METADATA

File: reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ignoreFields,
```
Comment (gorbunkov) [inline reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java:33]: Why do we need this "ignoreFields" strategy? Why to serialize an entity with no fields? As I see this enum value is never used in the code.

File: reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        convertAll
```
Comment (gorbunkov) [inline reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java:34]: convertAll is also never used. Why do we need it?

File: reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import com.google.gson.ExclusionStrategy;
+import com.google.gson.FieldAttributes;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.util.Arrays;
+
+public class StrategyGsonBuilder {
```
Comment (gorbunkov) [inline reports/src/main/java/io/jmix/reports/app/StrategyGsonBuilder.java:27]: A very common class name that doesn't explain what the class does. What strategy does it build? As I see, you use only ignoreMetadata serialization policy. Maybe it's worth thinking about removing the SerializationPolicy enum at all and rename the class into something like `MetadataIgnoringGsonBuilder`?

**Review by gorbunkov — CHANGES_REQUESTED** [review]

**Review by gorbunkov — APPROVED** [review]

**Pull request merged**